chain/chain/src/flat_storage_resharder.rs

```diff
@@ -314,63 +310,111 @@ impl FlatStorageResharder {
                 return FlatStorageReshardingTaskResult::Failed;
             }
-            FlatStorageReshardingTaskSchedulingStatus::Postponed => {
-                info!(target: "resharding", "flat storage shard split task has been postponed");
-                return FlatStorageReshardingTaskResult::Postponed;
-            }
         };
 
-        #[cfg(feature = "test_features")]
-        {
-            if self.adv_should_delay_task(&resharding_hash, chain_store) {
-                info!(target: "resharding", "flat storage shard split task has been artificially postponed!");
-                return FlatStorageReshardingTaskResult::Postponed;
-            }
-        }
-
-        // We know that the resharding block has become final so let's start the real work.
-        let (parent_shard, split_params) = self
-            .get_parent_shard_and_split_params()
-            .expect("flat storage resharding event must be Split!");
         let metrics = FlatStorageReshardingShardSplitMetrics::new(
             parent_shard,
             split_params.left_child_shard,
             split_params.right_child_shard,
         );
 
-        let task_status = self.split_shard_task_impl(parent_shard, &split_params, &metrics);
-        self.split_shard_task_postprocessing(parent_shard, split_params, &metrics, task_status);
+        let task_status =
+            self.split_shard_task_impl(parent_shard, &split_params, &resharding_block, &metrics);
+        self.split_shard_task_postprocessing(
+            parent_shard,
+            split_params,
+            &resharding_block,
+            &metrics,
+            task_status,
+        );
         info!(target: "resharding", ?task_status, "flat storage shard split task finished");
         task_status
     }
 
-    /// Performs the bulk of [split_shard_task].
+    /// Checks if the resharder is ready to split a shard. In such case it returns the parameters needed for the split.
+    fn split_shard_check_readiness(&self, chain_store: &ChainStore) -> SplitShardSchedulingStatus {
+        info!(target: "resharding", "checking readiness of flat storage shard split task");
+        // The current resharding event must exist and be compatible with the operation of
+        // splitting a shard. All these checks must happen while the resharding event is under
+        // lock, to prevent multiple parallel resharding attempts (due to forks) to interfere
+        // with each other .
+        let mut event_lock_guard = self.resharding_event.lock().unwrap();
+        let (resharding_blocks, parent_shard, split_params, execution_status) = if let Some(event) =
+            event_lock_guard.as_mut()
+        {
+            if event.has_started() {
+                info!(target: "resharding", "flat storage shard split task cancelled: resharding already in progress");
+                return SplitShardSchedulingStatus::Cancelled;
+            }
```

Comment:
Is this expected to happen under normal circumstances, under fork or under error?

Reply:
Only if there are forks involved